libs/autotagging-commons/src/main/java/org/opensearch/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,215 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.opensearch.autotagging.Rule._ID_STRING;
+
+/**
+ * This class encapsulates the logic to manage the lifecycle of rules at index level
+ * @opensearch.experimental
+ */
+public abstract class RulePersistenceService {
+    /**
+     * The system index name used for storing rules
+     */
+    public static final String RULES_INDEX = ".rules";
+    private final Client client;
+    private final ClusterService clusterService;
+    private static final Logger logger = LogManager.getLogger(RulePersistenceService.class);
+
+    /**
+     * The maximum number of results allowed per GET request
+     */
+    public static final int MAX_RETURN_SIZE_ALLOWED_PER_GET_REQUEST = 50;
+
+    /**
+     * Constructor for RulePersistenceService
+     * @param clusterService - The {@link ClusterService} to be used by RulePersistenceService
+     * @param client - The {@link Client} to be used by RulePersistenceService
+     */
+    @Inject
+    public RulePersistenceService(final ClusterService clusterService, final Client client) {
+        this.clusterService = clusterService;
+        this.client = client;
+    }
+
+    /**
+     * Entry point for the get rule api logic in persistence service. If id is provided, we only get a single rule.
+     * Otherwise, we get all rules that satisfy the attributeFilters.
+     * @param id - The id of the rule to get.
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param searchAfter - The sort values from the last document of the previous page, used for pagination
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    public void getRule(
+        String id,
+        Map<Attribute, Set<String>> attributeFilters,
+        String searchAfter,
+        ActionListener<? extends GetRuleResponse> listener
+    ) {
+        // Stash the current thread context when interacting with system index to perform
+        // operations as the system itself, bypassing authorization checks. This ensures that
+        // actions within this block are trusted and executed with system-level privileges.
+        try (ThreadContext.StoredContext context = getContext()) {
+            BoolQueryBuilder boolQuery = buildGetRuleQuery(id, attributeFilters);
+            SearchRequestBuilder searchRequest = client.prepareSearch(RULES_INDEX)
+                .setQuery(boolQuery)
+                .setSize(getMaxReturnSizeAllowedPerGetRequest());
+            if (searchAfter != null) {
+                searchRequest.addSort(_ID_STRING, SortOrder.ASC).searchAfter(new Object[] { searchAfter });
+            }
+            searchRequest.execute(ActionListener.wrap(searchResponse -> handleGetRuleResponse(id, searchResponse, listener), e -> {
+                logger.error("Failed to fetch all rules: {}", e.getMessage());
+                listener.onFailure(e);
+            }));
+        }
```

Comment:
given features will use this class generically, we should not tie the storage to a system index for all client features for this lib but rather provide with an interface which features can implement and use inside their `TransportActions`